securesystemslib/signer.py

```diff
@@ -0,0 +1,137 @@
+from typing import Dict
+
+
+class Signature:
+    """Storage class containing information about a signature and the keyid
+        uniquely identifying the key used to generate the signature.
+
+    Attributes:
+        keyid: HEX string used as a unique identifier of the key.
+        signature: HEX string representing the signature.
+    """
+    def __init__(self, keyid: str, sig: str):
+        self.keyid = keyid
+        self.signature = sig
+
+
+    @classmethod
+    def from_dict(cls, signature_dict: Dict) -> 'Signature':
+        """Creates a Signature object from its JSON/dict representation.
+
+        Arguments:
+            signature_dict: A dict that contains a valid keyid and signature.
+            Note that the fields should be named "keyid" and "sig" respectively.
+
+        Returns:
+            A "Signature" instance.
+        """
+
+        return cls(signature_dict["keyid"], signature_dict["sig"])
+
+
+    def to_dict(self) -> Dict:
+        """Returns the JSON-serializable dictionary representation of self."""
+
+        return {
+            "keyid": self.keyid,
+            "sig": self.signature
+        }
+
+
+
+class Signer:
+    """Signer interface created to support multiple signing implementations."""
+
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def sign(payload: bytes) -> 'Signature':
+        """
+        Purpose:
+            Abstract function used for signing a given payload by the key
+            assigned to the Signer instance.
+
+        Arguments:
+            payload: bytes to be signed
+
+        Returns:
+            Returns a "Signature" class instance.
+        """
+        raise NotImplementedError # pragma: no cover
+
+
+
+class SSlibSigner(Signer):
+    """
+    Purpose:
+        Securesystemslib default implementation of the "Signer" interface.
+        With this implementation the following signature schemes are supported:
+
+        'rsassa-pss-(md5|sha1|sha224|sha256|sha384|sha512)'
+        'rsa-pkcs1v15-(md5|sha1|sha224|sha256|sha384|sha512)'
+        RFC3447 - RSASSA-PSS
+        http://www.ietf.org/rfc/rfc3447.
+
+        'ed25519'
+        ed25519 - high-speed high security signatures
+        http://ed25519.cr.yp.to/
+
+        'ecdsa-sha2-nistp256'
+        https://tools.ietf.org/html/rfc5656
+
+    Attributes:
+        key_dict:
+            A dictionary containing the keys. Both private and public keys are
+            needed.
+            Which signature to generate is determined by the key type of
+            'key_dict'.
+
+            An example RSA key dict has the form:
+
+            {'keytype': 'rsa',
+            'scheme': 'rsassa-pss-sha256',
+            'keyid': 'f30a0870d026980100c0573bd557394f8c1bbd6...',
+            'keyval': {'public': '-----BEGIN RSA PUBLIC KEY----- ...',
+                        'private': '-----BEGIN RSA PRIVATE KEY----- ...'}}
+
+            The public and private keys are strings in PEM format.
+    """
```

Comment:
I tried to clarify the parts about key types and schemes a bit and ended up re-writing the entire docstring. What do you think?

```suggestion
    """A securesystemslib signer implementation.

    Provides a sign method to generate a cryptographic signature with a
    securesystemslib-style rsa, ed25519 or ecdsa private key on the instance.
    The signature scheme is determined by the key and must be one of:

    - rsa(ssa-pss|pkcs1v15)-(md5|sha1|sha224|sha256|sha384|sha512) (12 schemes)
    - ed25519
    - ecdsa-sha2-nistp256

    See 'securesystemslib.interface' for functions to generate and load keys.

    Attributes:
        key_dict:
            A securesystemslib-style key dictionary, which includes a keyid,
            key type, signature scheme, and the public and private key values,
            e.g.::

                {
                    "keytype": "rsa",
                    "scheme": "rsassa-pss-sha256",
                    "keyid": "f30a0870d026980100c0573bd557394f8c1bbd6...",
                    "keyval": {
                        "public": "-----BEGIN RSA PUBLIC KEY----- ...",
                        "private": "-----BEGIN RSA PRIVATE KEY----- ..."
                    }
                }

    """
```

Note that the code block now also renders nicely with sphinx.

Reply:
Looks great. I will use your suggestion.